library/core/src/primitive_docs.rs

```diff
@@ -1527,6 +1527,46 @@ mod prim_ref {}
 /// Note that all of this is not portable to platforms where function pointers and data pointers
 /// have different sizes.
 ///
+/// ### ABI compatibility
+///
+/// Generally, when a function is declared with one signature and called via a function pointer
+/// with a different signature, the two signatures must be *ABI-compatible* or else this call is
+/// Undefined Behavior. ABI compatibility is a lot stricter than merely having the same
+/// representation in memory; for example, even if `i32` and `f32` have the same size and alignment,
+/// they might be passed in different registers and hence not be ABI-compatible.
+///
+/// For two signatures to be considered *ABI-compatible*, they must declare the same `extern` ABI
+/// string, must take the same number of arguments, and the individual argument types and the return
+/// types must be ABI-compatible.
+/// The relation of when two types are ABI-compatible is defined as follows:
+///
+/// - Every type is ABI-compatible with itself.
+/// - If `<T as Pointee>::Metadata == ()`, then `*const T`, `*mut T`, `&T`, `&mut T`, `Box<T>`,
+///   `NonNull<T>` are all ABI-compatible with each other.
+/// - Any two `fn()` types with the same `extern` ABI string are ABI-compatible with each other.
+/// - Any two 1-ZST types (types with size 0 and alignment 1) are ABI-compatible.
+/// - A `repr(transparent)` type `T` is ABI-compatible with its unique non-1-ZST field (if there is
+///   such a field).
+/// - `i32` is ABI-compatible with `NonZeroI32`, and similar for all other integer types with their
+///   matching `NonZero*` type.
+/// - If `T` is guaranteed to be subject to the [null pointer
+///   optimization](option/index.html#representation), then `T` and `Option<T>` are ABI-compatible.
+/// - If `T1` and `T2` are ABI-compatible, then two `repr(C)` types that only differ because one
+///   field type was changed from `T1` to `T2` are ABI-compatible.
+/// - ABI-compatibility is symmetric and transitive.
+///
+/// Noteworthy cases of types *not* being ABI-compatible are `bool` vs `u8`, and `i32` vs `u32`: on
+/// some targets, the calling conventions for these types differ in terms of what they guarantee for
```

Comment:
It should probably not be undefined behavior on all targets to call `fn(i32)` with `fn(u32)`. For example with `extern "C"` on x86-64, these two types are ABI compatible so it should be fine and not undefined behavior to do that. It makes code less portable and should probably be flagged by Miri but I'm not sure whether its useful to say that it's always UB.
Maybe this could be handwaved by saying "on some targets with some ABIs, some types not listed here are ABI compatible as well".

Reply:
I don't think they are abi compatible on x86_64. If I recall correctly there used to be an abi incompatibility between gcc and clang precisely because of that.

Reply:
At least according to the `PassMode`, they are ABI-compatible on x86_64, but not on s390x.

Reply:
> In my opinion it should not be undefined behavior on all targets to call extern "C" fn(i32) with extern "C" fn(u32). 

Why not? Does the C standard permit such calls?

> Maybe this could be handwaved by saying "on some targets with some ABIs, some types not listed here are ABI compatible as well".

We'd have to add "but this can change any time and no stability guarantees are made", which makes this fairly useless.

Reply:
The text now says that "More signatures can be ABI-compatible on specific targets, but that should not be relied upon since it is not portable and not a stable guarantee", this also covers the i32 vs u32 case.